File: lms-studio/schemaTypes/attendance.ts
import { defineType, defineField } from 'sanity'

export const attendance = defineType({
  name: 'attendance',
  title: 'Attendance',
  type: 'document',
  fields: [
    defineField({
      name: 'student',
      title: 'Student',
      type: 'reference',
      to: [{ type: 'student' }],
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'date',
      title: 'Date',
      type: 'date',
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'status',
      title: 'Status',
      type: 'string',
      options: {
        list: [
          { title: 'Present', value: 'present' },
          { title: 'Absent', value: 'absent' },
          { title: 'Late', value: 'late' },
          { title: 'Excused', value: 'excused' },
        ],
      },
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'subject',
      title: 'Subject',
      type: 'reference',
      to: [{ type: 'subject' }],
    }),
    defineField({
      name: 'remarks',
      title: 'Remarks',
      type: 'string',
    }),
    defineField({
      name: 'markedBy',
      title: 'Marked By',
      type: 'reference',
      to: [{ type: 'teacher' }],
    }),
  ],
  preview: {
    select: {
      title: 'student.name',
      subtitle: 'date',
      status: 'status',
    },
    prepare({ title, subtitle, status }) {
      const emoji =
        status === 'present'
          ? '✅'
          : status === 'absent'
          ? '❌'
          : '⏰'
      return {
        title: `${emoji} ${title || 'Unknown'}`,
        subtitle,
      }
    },
  },
})

File: l_m_s/features/lms/feature_lms/lib/models/attendance.dart
/// Domain model for attendance (Sanity attendance type).
class AttendanceRecord {
  final String id;
  final String studentId;
  final String? studentName;
  final DateTime? date;
  final String status; // present, absent, late, excused
  final String? remarks;
  final String? subjectId;
  final String? batchId;

  const AttendanceRecord({
    required this.id,
    required this.studentId,
    this.studentName,
    this.date,
    this.status = 'present',
    this.remarks,
    this.subjectId,
    this.batchId,
  });

  factory AttendanceRecord.fromMap(Map<String, dynamic> map) {
    String studentId = '';
    if (map['student'] is Map) studentId = (map['student']['_ref'] ?? map['student']['_id'] ?? '').toString();
    if (studentId.isEmpty) studentId = map['studentId'] as String? ?? '';
    String? studentName;
    if (map['student'] is Map && map['student']['name'] != null) studentName = map['student']['name'] as String?;

    DateTime? date;
    final d = map['date'];
    if (d != null) {
      if (d is DateTime) {
        date = d;
      } else if (d is String) date = DateTime.tryParse(d);
    }

    return AttendanceRecord(
      id: map['_id'] as String? ?? '',
      studentId: studentId,
      studentName: studentName,
      date: date,
      status: map['status'] as String? ?? 'present',
      remarks: map['remarks'] as String?,
      subjectId: map['subject'] is Map ? (map['subject']['_ref'] ?? map['subject']['_id']) as String? : map['subjectId'] as String?,
      batchId: map['batchId'] as String?,
    );
  }
}
